src/ZodError.ts

```diff
@@ -123,15 +123,14 @@ export const quotelessJson = (obj: any) => {
   return json.replace(/"([^"]+)":/g, "$1:");
 };
 
-export class ZodError extends Error {
+export class ZodError {
```

Comment:
`ZodError` has to extend `Error` in order to preserve stack trace information.

Reply:
...also so it can be caught by users naively using `instanceof Error`.

Reply:
This is the core performance issue. Zod overuses try/catch in flow. Getting stack information is costly. Removing `Error` is quick fix. I guess the best way would be to change flow so it wouldn't be overused

Reply:
>I guess the best way would be to change flow so it wouldn't be overused

I fully agree with you. And in fact it looks like your other changes are going a long way in that direction?

Reply:
Well I tried to remove as many error creations as I could. 
Performance issues in zod are really making me worried. Even our requests where we validate data goes up from 300ms to 1600ms.

Reply:
I share your concern. We use Zod in production too and if possible we would like to drive its performance up drastically, at least in the happy path. However, removing the inheritance of `Error` would be a breaking chance that I doubt would be palatable to @colinhacks, never mind the users.

In lieu of this change, there are a few things we could do to ensure that `ZodError` is instantiated as rarely as possible. You've already done some of these things in this PR, which I'm thankful for.

My intuition is that the next step after merging your changes to `src/parser.ts` should be to run benchmarks and look at flame graphs, but this isn't something I have much experience with, nor the time to perform right now. Help is always appreciated. While we don't have benchmarks we can also look at making non-API-breaking optimizations.

Reply:
I zoomed out and saw that you had benchmarks above, which is wonderful! Would you be able to run another benchmark with just your changes to `src/parser.ts`?

Reply:
That what I was doing at first. As it does help with types where there are no unions. It does almost nothing where they are. Every union fail creates and throws an error :/ 
That's why my benchmarks only improved when I got rid of `Error`

Reply:
Oh wow, I didn't realize that was the problem. indeed, union parsing should collect issues without instantiating `Error`.

Let's give @colinhacks a few days to chime in in case there's a particular approach he'd like to take. I understand he's busy with work at the moment.